### PR TITLE
Account Protection: fix compatibility with stdClass in password-related hooks

### PR DESCRIPTION
**Pull request by Initsogar**
Fixes p1744318146887419/1744318128.196479-slack-C080AF0NF0R

## Proposed changes:
* Update Password_Manager to properly handle stdClass objects in WordPress hooks
* Add validation for user_pass property before saving password hashes
* Skip saving empty password hashes
* Add test coverage for all edge cases

The changes address compatibility issues with third-party code that passes stdClass objects instead of WP_User objects to WordPress hooks like profile_update and after_password_reset.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p1744318146887419/1744318128.196479-slack-C080AF0NF0R

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* Run unit tests and ensure it passes
```
cd projects/packages/account-protection
composer phpunit
```

### CODE REVIEW

File: projects/packages/account-protection/src/class-password-manager.php [hunk 1]
```diff
@@ -95,26 +95,33 @@ public function validate_password_reset( \WP_Error $errors, $user ): void {
...
 	 * Handle the password reset.
 	 *
-	 * @param \WP_User $user The user.
+	 * @param \WP_User|\stdClass $user The user object.
 	 *
 	 * @return void
 	 */
-	public function on_password_reset( \WP_User $user ): void {
+	public function on_password_reset( $user ): void {
+		if ( ! isset( $user->ID ) || ! isset( $user->user_pass ) || empty( $user->user_pass ) ) {
```
Comment (tbradsha) [inline projects/packages/account-protection/src/class-password-manager.php]: Could just do this if you want:
```suggestion
		if ( ! isset( $user->ID ) || empty( $user->user_pass ) ) {
```

Either way this won't save an old password of `0`. Not sure if we care.

Just to prevent inconsistencies down the road, the password check condition in `on_profile_update` vs. `on_password_reset` should probably be identical. Better yet, any reason to not just move this check directly into `save_recent_password_hash`?

**Review by tbradsha — COMMENTED** [review]
One comment inline.

Thanks for adding more tests! Tests pass locally and in CI, and overall this looks good to me, but I'd prefer someone with more context to give the green light.

File: projects/packages/account-protection/src/class-password-manager.php [hunk 2]
```diff
@@ -95,26 +95,33 @@ public function validate_password_reset( \WP_Error $errors, $user ): void {
...
 	 *
 	 * @return void
 	 */
-	public function on_profile_update( int $user_id, \WP_User $old_user_data ): void {
+	public function on_profile_update( int $user_id, $old_user_data ): void {
 		// phpcs:ignore WordPress.Security.NonceVerification
 		if ( isset( $_POST['action'] ) && $_POST['action'] === 'update' ) {
-			$this->save_recent_password_hash( $user_id, $old_user_data->user_pass );
+			// Only save if user_pass property exists and is not empty
+			if ( isset( $old_user_data->user_pass ) && '' !== $old_user_data->user_pass ) {
```
Comment (macbre) [inline projects/packages/account-protection/src/class-password-manager.php]:
```suggestion
			if ( isset( $old_user_data->user_pass ?? null ) && '' !== $old_user_data->user_pass ) {
```

We've relaxed the type checking for the `old_user_data` argument, let's stay on the safe side :-)

File: projects/packages/account-protection/tests/php/Password_Manager_Test.php [hunk 3]
```diff
@@ -88,24 +93,108 @@ public function test_on_profile_update_with_valid_nonce() {
...
+		$password_manager_mock->expects( $this->never() )
+			->method( 'save_recent_password_hash' );
+
+		$password_manager_mock->on_password_reset( $user );
+	}
+
+	public function test_on_password_reset_with_empty_user_pass() {
+		$user            = new \stdClass();
+		$user->ID        = 1;
+		$user->user_pass = '';
```
Comment (macbre) [inline projects/packages/account-protection/tests/php/Password_Manager_Test.php:199]: Can we add a test case where non-object (say a `null` value) is passed?

Comment (Initsogar) [reply projects/packages/account-protection/tests/php/Password_Manager_Test.php:199]: Done! https://github.com/Automattic/jetpack/pull/43045/commits/176a5160679a76699d257710a0ebfb29ed6953cc

Comment (macbre) [reply projects/packages/account-protection/tests/php/Password_Manager_Test.php:199]: Thanks!

File: projects/packages/account-protection/src/class-password-manager.php [hunk 4]
```diff
@@ -95,12 +95,16 @@ public function validate_password_reset( \WP_Error $errors, $user ): void {
...
-	 * @param int      $user_id The user ID.
-	 * @param \WP_User $old_user_data Object containing user data prior to update.
+	 * @param int                $user_id The user ID.
+	 * @param \WP_User|\stdClass $old_user_data Object containing user data prior to update.
 	 *
 	 * @return void
 	 */
-	public function on_profile_update( int $user_id, \WP_User $old_user_data ): void {
+	public function on_profile_update( int $user_id, $old_user_data ): void {
+		if ( ! is_object( $old_user_data ) || ! isset( $old_user_data->user_pass ) || '' === $old_user_data->user_pass ) {
```
Comment (Initsogar) [inline projects/packages/account-protection/src/class-password-manager.php]: I just refactored a bit to use a guard clause and added an is_object() clause as well. Thoughts? @macbre

File: projects/packages/account-protection/src/class-password-manager.php (hunk 1 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-manager.php]: > Either way this won't save an old password of 0. Not sure if we care.

Since I'm not 100% familiar with the feature, I would prefer to keep it as much explicit as possible.
I updated the guard clause to also consider if it is an object:

https://github.com/Automattic/jetpack/pull/43045/files#diff-fe8256f355e8848c15a4c61a610b3b7c0ef832d25bdc0814547792476abfb04bR122

> Just to prevent inconsistencies down the road, the password check condition in on_profile_update vs. on_password_reset should probably be identical. Better yet, any reason to not just move this check directly into save_recent_password_hash?

This suggestion makes sense. I'd prefer to keep it separate for now, as I have some unmerged PRs related to this feature that I'd like to review before and avoid any potential code conflicts.

@tbradsha would you be okay by merging this and monitor it is all good and revisit this in a different PR?

Comment (tbradsha) [reply projects/packages/account-protection/src/class-password-manager.php]: > I would prefer to keep it as much explicit as possible.

I'll note that it's identical logic with or without the `isset()`.

> I'd prefer to keep it separate for now

👍 

> would you be okay by merging this

Yes, I don't think we've broken things, and even if we missed an edge case it should prevent several failures. It's not clear if you understood my comment about a `0` password; though. Prior to this PR, I believe it would log the old password hash if the old password was falsy (i.e. `0`). Now it will not. I don't think it matters, but it is a change this PR introduces.

File: projects/packages/account-protection/src/class-password-manager.php (hunk 4 above)
Comment (macbre) [reply projects/packages/account-protection/src/class-password-manager.php]: I've became a huge fan of the `??` operator when it got introduced in PHP :-) But `is_object` should do it as well.

One could even simplify that condition to:

```suggestion
		if ( '' === ( $old_user_data->user_pass ?? '' ) ) {
```

```php
> $old_user_data = null;
> var_dump( $old_user_data->user_pass ?? '' );
string(0) ""
```

Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-manager.php]: Gotcha. I think your approach will also require an `is_object()` check, since property access happens before null coalescing. Even if the method isn’t expected to receive anything other than an object, I can’t control what other plugins or themes might do, so I still need to ensure we’re receiving an object that includes the `user_pass` property.

With that in mind, I simplified the guard clause a bit by using `is_object()` along with `empty()`: https://github.com/Automattic/jetpack/pull/43045/commits/2a437ce671302a6921569adfdde34a80480abfd5

Thanks for the tip! The `??` operator is something I’d love to play around with more in PHP :)

File: projects/packages/account-protection/src/class-password-manager.php (hunk 1 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-manager.php]: > Yes, I don't think we've broken things, and even if we missed an edge case it should prevent several failures. It's not clear if you understood my comment about a 0 password; though. Prior to this PR, I believe it would log the old password hash if the old password was falsy (i.e. 0). Now it will not. I don't think it matters, but it is a change this PR introduces.

Thanks for the clarification! I don’t think there will be an issue with `0` either. I went ahead with `is_object()` and `empty()` to simplify things while keeping it safe 🙂. Let me know what you think! https://github.com/Automattic/jetpack/pull/43045/commits/2a437ce671302a6921569adfdde34a80480abfd5

File: projects/packages/account-protection/src/class-password-manager.php (hunk 4 above)
Comment (macbre) [reply projects/packages/account-protection/src/class-password-manager.php]: > I think your approach will also require an `is_object()` check, since property access happens before null coalescing

Not really, that's why the null coalescing operator is so handy:

```php
> $old_user_data = null;
> var_dump( $old_user_data->user_pass ?? '' );
string(0) ""
```

**Review by tbradsha — APPROVED** [review]

File: projects/packages/account-protection/src/class-password-manager.php (hunk 4 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-manager.php]: Hmmm, I think something else was broken when I tested it then 🥲.
Sounds good, I'll keep it in mind next time, thanks!
